provisioning/service/src/Authentication/ServiceConnectionString.cs

```diff
@@ -1,156 +1,235 @@
 ﻿// Copyright (c) Microsoft. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
+using Microsoft.Azure.Devices.Common;
 using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq.Expressions;
 using System.Text;
+using System.Text.RegularExpressions;
+using SharedAccessSignatureParser = Microsoft.Azure.Devices.Provisioning.Service.SharedAccessSignature;
 
 namespace Microsoft.Azure.Devices.Provisioning.Service
 {
     /// <summary>
-    /// This object handles the connection string for the Azure IoT Services.
+    /// The Service Connection String Parser class
     /// </summary>
-    /// <remarks>
-    /// The connection string contains a set of information that uniquely identify an IoT Service.
-    /// 
-    /// A valid connection string shall be in one of the following formats:
-    /// <code>
-    /// HostName=[ServiceName];SharedAccessKeyName=[keyName];SharedAccessKey=[Key]
-    /// HostName=[ServiceName];SharedAccessKeyName=[keyName];SharedAccessSignature=[Signature]
-    /// </code>
-    /// 
-    /// This object parse and store the connection string. It is responsible to provide the authorization token too. 
-    /// </remarks>
-    public sealed class ServiceConnectionString
+    public class ServiceConnectionString
     {
-        private static readonly TimeSpan DefaultTokenTimeToLive = TimeSpan.FromMinutes(5);
+        private const char ValuePairDelimiter = ';';
+        private const char ValuePairSeparator = '=';
+        private const string HostNameSeparator = ".";
+
+        private const string HostNamePropertyName = nameof(HostName);
+        private const string SharedAccessKeyNamePropertyName = nameof(SharedAccessKeyName);
+        private const string SharedAccessKeyPropertyName = nameof(SharedAccessKey);
+        private const string SharedAccessSignaturePropertyName = nameof(SharedAccessSignature);
+
+        private static readonly TimeSpan regexTimeoutMilliseconds = TimeSpan.FromMilliseconds(500);
+        private static readonly Regex HostNameRegex = new Regex(@"[a-zA-Z0-9_\-\.]+$", RegexOptions.Compiled | RegexOptions.IgnoreCase, regexTimeoutMilliseconds);
```

Comment:
Please look at the memory profile - we're wasting a lot of memory compiling 4 Regex objects. 
Is there any time when we would only need to run one but not the others? If no, let's write a single RegEx (or, even better, our own string parser) that captures all elements.